src/channels/cave-story/index.tsx

```diff
@@ -0,0 +1,403 @@
+import type { FormattedDonation, Total } from '@gdq/types/tracker';
+import { ChannelProps, registerChannel } from '../channels';
+
+import { useListenFor, useReplicant } from 'use-nodecg';
+import styled from '@emotion/styled';
+import { css, keyframes } from '@emotion/react';
+import TweenNumber from '@gdq/lib/components/TweenNumber';
+
+import { useEffect, useRef } from 'react';
+import React from 'react';
+import { useRafLoop } from 'react-use';
+
+import clouds1 from './images/clouds1.png';
+import clouds2 from './images/clouds2.png';
+import clouds3 from './images/clouds3.png';
+import clouds4 from './images/clouds4.png';
+import clouds5 from './images/clouds5.png';
+import dragon from './images/dragon.gif';
+import sue from './images/sue.gif';
+import level from './images/level.png';
+import balrog from './images/balrog.gif';
+import ghostcat from './images/ghostcat.gif';
+import bat from './images/bat.gif';
+import critterSmall from './images/critterSmall.gif';
+import critterBig from './images/critterBig.gif';
+import misery from './images/misery.gif';
+import puppy from './images/puppy.gif';
+import quote from './images/quote.gif';
+import quoteMask from './images/quoteMask.gif';
+import curly from './images/curly.gif';
+
+registerChannel('Cave Story', 290, CaveStory, {
+	position: 'bottomLeft',
+	site: 'GitHub',
+	handle: 'omgitsraven',
+});
+
+function CaveStory(props: ChannelProps) {
+	const [total] = useReplicant<Total | null>('total', null);
+
+	const flyingDonationsHolder = useRef<FlyingDonationsHolder>(null);
+
+	var curDeck: Array<string> = [];
+	var nextDeck: Array<string> = [];
+	function putIntoBackHalfOfDeck(item:string){
+		const startHalfLength = Math.ceil(nextDeck.length/2);
+		const remainingHalfLength = nextDeck.length-startHalfLength;
+		const finalIndex = startHalfLength + Math.floor(Math.random()*(remainingHalfLength+1));
+		nextDeck.splice(finalIndex, 0, item);
+	}
+	putIntoBackHalfOfDeck("Misery");
+	putIntoBackHalfOfDeck("Balrog");
+	putIntoBackHalfOfDeck("Quote");
+	putIntoBackHalfOfDeck("Curly");
+	putIntoBackHalfOfDeck("Puppy");
+	curDeck = nextDeck;
+	nextDeck = [];
```

Comment:
This will be executed every time the component is re-rendered, constructing new arrays each time. However, only one of these sets of arrays is used because they are captured inside of the later closures. These should be put inside refs and use a ``useEffect``.

Reply:
thank you for the feedback;  I wound up realising I was really misunderstanding how React works, heh.  After a few more commits, this new version should now:
1) use refs for any data that persists, and
2) use a functional component rather than a class (for the only remaining component).
(useEffect wound up not being necessary because I put all the setup logic into the useListenFor hook instead.)
Also while I was testing I made a cosmetic tweak to the centering on the flying characters.
Hopefully this should suit better, please let me know if there's anything I'm still overlooking! Thanks!